### DISCUSSION TITLE
Design Idea: Command and Control, Organizational Leadership, Election

### DISCUSSION THREAD

**Discussion by mokun** [mars-sim/mars-sim#428] [Ideas]
As of r4959, each settler is assigned one job and one role either in the xml or by the algorithm at the start of the sim.

While the role is indicative of one's hierarchical position in a settlement, the job dictates what kind of tasks to perform. 

Player may request changing the job of a settler during the simulation. The final approval is  

Currently, there is only ONE rule book for setting up the leadership structure, as explained in the [Role wiki](https://github.com/mars-sim/mars-sim/wiki/Role). e.g. 3 types of heads, 7 types of division chiefs and specialists, etc.

For now, the top echelon roles are called `Commander`, `Sub-Commander`, `Chiefs` and `Mayor', etc.. They are elected primarily based on their attributes and skillset. 

We envision the rules governing the `Command and Control (C2)` of a settlement be more flexible than following just pre-defined rules.

The [ChainOfCommand](https://github.com/mars-sim/mars-sim/blob/e5c373e6ee31899cf6b80111b94cba775ac8c0d4/mars-sim-core/src/main/java/org/mars_sim/msp/core/structure/ChainOfCommand.java#L352) class has been the Jack-of-all-trade class responsible for tracking and reassigning roles and handling the election. 

```java
	/**
	 * Assigns a person with a specialist roleType in 7-division settlement
	 * 
	 * @param person {@link Person}
	 */
	public void assignSpecialiststo7Divisions(Person person) {
```

The rigid branching conditions in the method [assignSpecialiststo7Divisions()](https://github.com/mars-sim/mars-sim/blob/master/mars-sim-core/src/main/java/org/mars_sim/msp/core/structure/ChainOfCommand.java#L352) are also less than optimal. It sometimes designates too many `Resource Specialists` while having just one `Logistic Specialist`.

Q1: How would you like to rewrite the algorithm in the method `assignSpecialiststo7Divisions()` ?

Q2: How should we programmatically design the class and methods to allow multiple `templates` for players to choose from ?

Q3: What does it take to allow players to put together `custom` election rules, instead of fixed rules ?

**Comment by mokun** [comment]
> Q1: How would you like to rewrite the algorithm in the method assignSpecialiststo7Divisions() ?

For role election, we may rewrite it in such a way that's similar to how a settler is elected to a particular job.  

https://github.com/mars-sim/mars-sim/blob/e5c373e6ee31899cf6b80111b94cba775ac8c0d4/mars-sim-core/src/main/java/org/mars_sim/msp/core/person/ai/job/JobManager.java#L340-L365

As seen from the block of codes above, we use the method `getJobProspect(()` to do compute a `Job Prospect` score for each job job on each person.

And we also adjust for what the settlement needs at the moment. 

After that, we use the method `findBestFit()` to compare these scores across the settlement to figure out who has the highest score and assign that individual to that particular job.

https://github.com/mars-sim/mars-sim/blob/e5c373e6ee31899cf6b80111b94cba775ac8c0d4/mars-sim-core/src/main/java/org/mars_sim/msp/core/person/ai/job/JobManager.java#L272-L292

> Note: player can see the job prospect scores in the form of a text table of a settler in the chat box.

And we can throw in a fair amount of `probability` to the role election so that it doesn't always have the same result each time.

**Comment by Urwumpe** [comment]
My suggestion there:

For making a good SPI example, I started a module mars-sim-organisation. This should contain everything about organisation aspects. mars-sim-core should not even know that it exists, but mars-sim-organisation requires core (This should prevent circular dependencies). 

For integrating the organisation module into the core simulation, I wanted to use the Task system. First of all, I wanted to allow adding different "TaskGiver" functions to a person. Those should be defined by SPI in the organisation module. Next, I wanted to define tasks to work with organisations.  For example, assigning a person to a function could be handled as Task by a commander of the organisational unit containing the unit in question. 

For storing the knowledge about the organisational context of a Person, I wanted to extend the Mind class towards being able to store generic "Information Repositories", which are unique for a person and which are created by a factory object again provided to the core by SPI. For example, the "ORG" knowledge of a person could contain a list of organisation and function tuples ("In organisation o, the person p has the function f"). 

The advantage at this point: Core must not know anything about organisations. We could use the extension mechanisms by SPI for other facets of the simulation. Everything that defines organisations is contained in one module. And by using tasks, everything related to organisations would be done by Persons in the simulation, instead of some magic code. 

For including this into core, I wanted to extend the existing functions by factory classes - maybe we can later move everything to the same system and simply have a factory class that gets its data from a [Service Loader](https://docs.oracle.com/en/java/javase/12/docs/api/java.base/java/util/ServiceLoader.html) .

For getting the initial data into the module, I wanted to extend core towards providing a list of XML documents that define the initial organisation templates for starting a new simulation, but expects the organisation module to read them (So core does not need to know the class definitions of that module). A similar system would be needed for loading and saving the state of a simulation: Core would need to delegate this to the module - ideally, by having separate JSON files in a compressed archive, one for each entity providing by a module (For example, data stored in the mind of a person)


For making this visible in UI, I wanted to use a generic representation interface, which again could be provided by SPI from the mars-sim-organisation module, instead of using a dedicated UI extension there. This way, every UI just needs to know how to translate a representation into a view. The simplest way I see there is to simply translate an simulation object to JSON and provide some constant metadata about how it should look like. For example, an organisation structure could then be displayed as organigram, while the organisation and function assignments of a Person could be rendered as table. 

(You see, here I did not yet think as much about about how to implement this)

Using JSON would have the advantage, that adding a web UI would be pretty straight forward then. And by using some meta data about how to render the JSON, any UI supporting such a system could quickly display new modules without changes. We would just need to extend the UI then, when the standard for the format definitions in mars-sim-core changes. And the views would only need to be generated from the representation data in core during start-up, so we maybe have a slower start, but would be very flexible for extension.
